core/src/main/java/org/elasticsearch/env/Environment.java

```diff
@@ -83,7 +87,38 @@
     private final Path pidFile;
 
     /** Path to the temporary file directory used by the JDK */
-    private final Path tmpFile = PathUtils.get(System.getProperty("java.io.tmpdir"));
+    private static final Path systemTmpFile = PathUtils.get(System.getProperty("java.io.tmpdir"));
+
+    /**
+     * Path to the temporary file directory used by Elasticsearch.
+     * This will be a sub-directory that is created under the JDK temporary directory.
+     */
+    private static final Path tmpFile;
+
+    /**
+     * Creates a secure temporary directory whose contents can only be seen by the user
+     * that Elasticsearch is running as.  This is static so that only one such directory
+     * is created, even when many instances of {@link Environment} are created.
+     */
+    static {
```

Comment:
It feels like this is the wrong place to do this. I think we should do this in `Bootstrap` and then just pass a `Path tmpDir` to `InternalSettingsPreparer#prepareEnvironment` and make sure it's mandatory for all instances of environment that are not the single arg ctor. Btw. it feels like there are some sleeping bugs if this ctor is used in prod code.